nixos/modules/services/hardware/udev.nix

```diff
@@ -298,16 +330,46 @@ in
 
     boot.kernelParams = mkIf (!config.networking.usePredictableInterfaceNames) [ "net.ifnames=0" ];
 
-    boot.initrd.extraUdevRulesCommands = optionalString (cfg.initrdRules != "")
+    boot.initrd.extraUdevRulesCommands = optionalString (!config.boot.initrd.systemd.enable && cfg.initrdRules != "")
       ''
         cat <<'EOF' > $out/99-local.rules
         ${cfg.initrdRules}
         EOF
       '';
 
+    # Generate the udev rules for the initrd
+    boot.initrd.systemd.contents = {
+      "/etc/udev/rules.d".source = udevRulesFor {
+        name = "initrd-udev-rules";
+        initrdBin = config.boot.initrd.systemd.contents."/bin".source;
+        udevPackages = cfg.initrdPackages;
+        udevPath = config.boot.initrd.systemd.contents."/bin".source;
+        udev = config.boot.initrd.systemd.package;
+        systemd = config.boot.initrd.systemd.package;
+        binPackages = cfg.initrdBinPackages ++ [ config.boot.initrd.systemd.contents."/bin".source ];
+      };
+    };
+    # Add the binaries as well
+    boot.initrd.systemd.storePaths = map toString cfg.initrdBinPackages;
+    # We need lvm2 for the device-mapper rules and we also need to add our custom rules
+    services.udev.initrdPackages = [ pkgs.lvm2 ] ++ optional (cfg.initrdRules != "") (pkgs.writeTextFile {
```

Comment:
Was just wondering: Do we instead want to have a `boot.initrd.services.udev` namespace?

Reply:
Done